api/api.go

```diff
@@ -902,8 +902,20 @@ func parseQueryMeta(resp *http.Response, q *QueryMeta) error {
 
 // decodeBody is used to JSON decode a body
 func decodeBody(resp *http.Response, out interface{}) error {
-	dec := json.NewDecoder(resp.Body)
-	return dec.Decode(out)
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response body: %s", err)
+	}
+
+	if err := json.NewDecoder(bytes.NewReader(b)).Decode(out); err != nil {
+		// we got an invalid json response, just return the
+		// body content as an error.
+		// This happens when the Consul agent returns an error like:
+		// "No cluster leader"
+		return fmt.Errorf(string(b))
```

Comment:
I'll try to improve a bit further: output the HTTP code received.

What about taking the first line ONLY and limit the size of error? (in case you are not talking to a real consul agent, ie: behind a proxy)

An error of the following form might be useful:

"HTTP 500: No cluster leader" (or whatever the code/message received of course)

Reply:
Implemented your suggestions to limit error message to the first line of the response body and include HTTP response code.